Add additional queue choices and request dropping at Ambry Server

Arun-LinkedIn opened the pull request:
This PR is mostly copied from @cgtz work in this [Pull request](https://github.com/linkedin/ambry/commit/d19707ed62e24da08ca5e01b19aa94e79a3bdef5) which adds following changes at Ambry server. 

- Add support for dropping requests that have exceeded a queueing timeout.
- Add an implementation of the Adaptive-LIFO co-del queue described in https://queue.acm.org/detail.cfm?id=2839461 that supports request dropping and based on an HBase java implementation. However, keeping this implementation disabled until we do few more tests to confirm on default config parameters for co-del algorithm.

litingulfs commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +102,65 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      networkRequestBundle = networkRequestQueue.take();
+      NetworkRequest unqueuedRequest;
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        networkRequestBundle.getRequestsToDrop().add(unqueuedRequest);
```
It seems counter-intuitive that you're immediately adding new requests to the queue for requests that may get dropped. Why?

litingulfs reviewed (changes requested):
This is very nice, I love the idea of making Ambry more resilient by adaptively dropping requests!

I'm new to this part of the Ambry code, so I might have misunderstood what is going on here. However, it seems to me that replication related requests may be dropped as well as client requests - is that correct?

I didn't see any metrics related to dropping requests or switching between queue policies. Unless I missed something here, would it be possible to add metrics for this? I suspect that we may want to alert if we drop a large number of requests for a significant duration.

Finally, is there any way in which we could roll out this implementation based on a lix? Even if it is just the dropping of requests that is gated by the lix?

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/protocol/RequestAPI.java:
```diff
@@ -25,77 +25,88 @@ public interface RequestAPI {
    * @throws InterruptedException if request processing is interrupted.
    */
-  void handleRequests(NetworkRequest request) throws InterruptedException;
+  void handleRequests(NetworkRequest request, boolean shouldDrop) throws InterruptedException;
```
These interfaces look weird to me for two reasons.
1. Dropping request should be done in netty io thread, it shouldn't invoke request handling requests.
2. Even if we want to do this in request handling request, can we have a new method like handleDropppedRequest?

JingQianCloud replied:
+1
Not sure how easy or hard to reject the requests in netty io thread since it doesn't deal with AmbryRequest before. But if it's doable, we will be able to reject these request at the earliest convenience. It's more efficient.

And with either 1 or 2, it can avoid the massive interface change.

snalli replied:
Would the netty thread print out any logs indicating that requests are being rejected ? or we might be in a limbo.

tbingol replied:
I think this cannot be achieved in the netty io thread since the decision to drop a request cannot be reached during enqueuing. Requests can decay while they are in the queue.

OTOH, +1 on having a generic request drop routing rather than delaying it to specific request handlers, e.g., handlePutRequest.

Arun-LinkedIn replied:
Sure. Made changes to have a separate method in RequestAPI to drop requests instead of modifying all `handlexxxRequest` methods. Also, moved metrics also to a common place so that they can be reused for both 'handled' and 'dropped' requests.

Even though it is ideal to send responses 'unqueued' requests immediately in Netty Thread, I was thinking it might be better to have same request-response flow path (i.e. requests queued by Netty thread and dequeued, processed and handed over back to netty outbound queues by request handler thread) for both 'unqueued' and 'delayed' requests. It was also easier to have response handling code (like updating metrics, etc) in one place in that way.

JingQianCloud commented on ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java:
```diff
@@ -0,0 +1,169 @@
+ * @see <a href="http://queue.acm.org/detail.cfm?id=2839461">Fail at Scale paper</a>
+ * @see <a href="https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/AdaptiveLifoCoDelCallQueue.java">HBase implementation</a>
+ */
+public class AdaptiveLifoCoDelNetworkRequestQueue implements NetworkRequestQueue {
```
Not sure how well the parameters are tuned for the adaptive queue and if it's suitable for our case.
To my own opinion, probably we can just start with some simple LIFO. But it's just my 2 cents.

Arun-LinkedIn replied:
Sure. Yeah, true. I was also thinking if we can start with simple Fifo queue first and later do more tests on adaptive queue to tune the config parameters.

tbingol commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +102,65 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToDrop.getStartTimeInMs());
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
```
Log the errors requests?

Arun-LinkedIn replied:
Sure, done.

Arun-LinkedIn commented on ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java:
```diff
@@ -0,0 +1,169 @@
+    while (true) {
+      NetworkRequest nextRequest;
+      if (useLifoMode()) {
+        nextRequest = deque.pollLast();
```
Sure. Added log statement to print when mode changes in `useLifoMode` method. For metrics, actually added a set of metrics at higher level to track dropped request rate for various types of requests. Since changing of queue mode is an implementation detail, wondering if metric may not be needed here. What do you say?

Arun-LinkedIn commented on ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java:
```diff
@@ -0,0 +1,169 @@
+   * @return {@code true} if the queue should operate in LIFO mode.
+   */
+  private boolean useLifoMode() {
+    double loadFactor = (double) deque.size() / maxCapacity;
```
Sure. Done

Arun-LinkedIn commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -30,18 +36,34 @@
   public void sendRequest(NetworkRequest request) throws InterruptedException {
-    requestQueue.put(request);
+    if (!networkRequestQueue.offer(request)) {
+      logger.debug("Request queue is full, dropping incoming request: {}", request);
```
Sure. Updated as warning message.

mikeyang923 commented on ambry-api/src/main/java/com/github/ambry/config/NetworkConfig.java:
```diff
@@ -119,6 +122,35 @@ public class NetworkConfig {
+  /**
+   * The threshold over which we switch the request queue from FIFO to LIFO mode.
+   */
+  @Config("adaptive.lifo.queue.threshold")
+  @Default("0.7")
+  public final double adaptiveLifoQueueThreshold;
```
Is those `@Default `actually in-use/needed? Since i see we explicitly set the defaultValue in constructor like below:
` adaptiveLifoQueueThreshold = verifiableProperties.getDoubleInRange("adaptive.queue.lifo.threshold", 0.7, 0.0, 1.0);`

Arun-LinkedIn replied:
Will check once. I copied from other configs but like you said`@Default` annotation may not be in use.

JingQianCloud commented on ambry-network/src/main/java/com/github/ambry/network/FifoNetworkRequestQueue.java:
```diff
@@ -0,0 +1,83 @@
+    while ((nextRequest = queue.poll()) != null) {
+      if (needToDrop(nextRequest)) {
+        logger.warn("Request timed out waiting in queue, dropping it: {}", nextRequest);
+        requestsToDrop.add(nextRequest);
```
When "needToDrop" is true, shall we also need return a request to server?
Looks like we either return one request to serve or one request to drop.

But for the definition of NetworkRequestBundle, it can have 1 request to serve and multiple requests to drop.

Arun-LinkedIn replied:
Hmm, in this case, if we find that the request needs to be dropped, we are not actually breaking from loop and are continuing to iterate until we find a valid request or the queue is empty. So, the `NetworkRequestBundle` could return 1  request to serve and multiple requests to drop.

JingQianCloud commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -30,18 +36,34 @@
   /** Send a request to be handled, potentially blocking until there is room in the queue for the request */
   @Override
   public void sendRequest(NetworkRequest request) throws InterruptedException {
-    requestQueue.put(request);
```
Looks like we have two places to decide drop request.
1. NetworkRequestQueue has a blocking queue. It may drop request as well.
2. RequestResponseChannel has a unqueuedRequests queue if NetworkRequestQueue blocking queue is full. 

Is that possible to let RequestResponseChannel manage all the requests and decide the drop policy? Don't split it to two places. NetworkRequestQueue can have its own queue instead of the blocking queue.

Arun-LinkedIn replied:
Yeah, true. Like you said, it is probably good to have `unqueuedRequests` queue as well in `NetworkRequestQueue`. But, with the latest commit, since I am fetching unqueued requests separately from `RequestResponseChannel`, I thought I will leave the `unqueuedRequests` queue in RequestResponseChannel. Please let me know what you think.

JingQianCloud commented on ambry-network/src/main/java/com/github/ambry/network/SocketRequestResponseChannel.java:
```diff
@@ -145,7 +162,10 @@ public SocketRequestResponseChannel(int numProcessors, int queueSize) {
   @Override
   public void sendRequest(NetworkRequest request) throws InterruptedException {
-    requestQueue.put(request);
+    if (!networkRequestQueue.offer(request)) {
```
Same here. 
If NetworkRequestQueue hold all the requests, we don't need unqueuedRequest queue for each RequestResponseChannel implementation.

Arun-LinkedIn commented:
Just wanted to add this PR has 3 changes. 

1. Adding fifo and adaptive queues to support request dropping as originally described in description.
2. Added metrics to track drop rate for various types of requests like put, get, etc. Also, moved the updating of metrics (like put request rate, get request rate, etc) to a separate method since it was turning out to a lot of duplicate code
3. Added a background thread (request-dropper thread) to do function of only dropping requests to avoid memory build since request-handler threads may be busy and cannot catch up with drop rate.

Please let me know if you want to split 2 and 3 points into separate PRs.

justinlin-linkedin commented on ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java:
```diff
@@ -196,6 +196,77 @@ public void handleRequests(NetworkRequest networkRequest) throws InterruptedExce
+      long requestQueueTime = SystemTime.getInstance().milliseconds() - request.getStartTimeInMs();
+      RequestMetricsUpdater metricsUpdater = new RequestMetricsUpdater(requestQueueTime, 0, 0, 0, true);
+      switch (type) {
+        case PutRequest:
```
nit: we don't have to deserialize the entire put request, we can just get the correlation id and the client id. we can probably do this in another PR.

justinlin-linkedin commented on ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java:
```diff
@@ -1246,4 +1253,322 @@ private void applyDelete(MessageInfo messageInfo, ReplicateBlobRequest replicate
+  /**
+   * Used to update common metrics for the incoming request.
+   */
+  public class RequestMetricsUpdater implements RequestVisitor {
```
I don't think we need to use visitor pattern given that we already have lots of code branches that do a switch statement based on RequestType. It seems like the handleXXXRequest is more suitable to be converted to a visitor. But, anyway, it's fine to use it.

snalli replied:
+1, these visit methods are just adding more unnecessary code

tbingol commented on ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java:
```diff
@@ -196,6 +196,77 @@ public void handleRequests(NetworkRequest networkRequest) throws InterruptedExce
+      Response response;
+      long requestQueueTime = SystemTime.getInstance().milliseconds() - request.getStartTimeInMs();
+      RequestMetricsUpdater metricsUpdater = new RequestMetricsUpdater(requestQueueTime, 0, 0, 0, true);
+      switch (type) {
```
Thanks for creating a function to drop all request types. However, I was wondering if we could return a generic throttling error rather than a request-specific one. The reason is that we can use other frameworks for load shedding, such as Hodor, in the future. Such frameworks will be application-unaware and will likely drop the requests before they make it even this far.

snalli commented on ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java:
```diff
@@ -0,0 +1,187 @@
+    // Dequeue next request. If the request timed out, continue to dequeue until we find a valid request to serve
+    // or queue is empty.
+    NetworkRequest nextRequest;
+    while (true) {
```
while (deque.size() > 0)

snalli commented on ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java:
```diff
@@ -0,0 +1,187 @@
+      } else {
+        nextRequest = deque.pollFirst();
+      }
+      if (nextRequest == null) {
```
if the loops condition is on size, then we can get rid of this break ?

snalli commented on ambry-network/src/main/java/com/github/ambry/network/FifoNetworkRequestQueue.java:
```diff
@@ -0,0 +1,97 @@
+    // If there are no requests to drop and no requests to serve currently in the queue, block until a new request comes
+    // so we can give the consumer something to do.
+    if (requestToServe == null && requestsToDrop.isEmpty()) {
+      requestToServe = queue.take();
```
i don't understand what's happening here. This call blocks or returns immediately if the queue is empty ?

snalli commented on ambry-network/src/main/java/com/github/ambry/network/FifoNetworkRequestQueue.java:
```diff
@@ -0,0 +1,97 @@
+
+/**
+ * A FIFO queue to hold network requests. It Internally, it maintains two queues, a bounded blocking queue to hold immediate
+ * set of next requests that can be handled by Ambry and unbounded queue to hold overflow requests.
```
where is this unbounded queue ? is it requestsToDrop ? is it a good idea to have an unbounded queue ?

snalli commented on ambry-network/src/main/java/com/github/ambry/network/http2/AmbryNetworkRequestHandler.java:
```diff
@@ -41,7 +41,7 @@ public AmbryNetworkRequestHandler(RequestResponseChannel requestResponseChannel,
   @Override
   protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest msg) {
-
+    // I received a message
```
+1

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -30,18 +38,34 @@
   }
 
-  /** Send a request to be handled, potentially blocking until there is room in the queue for the request */
+  /** Send a request to be handled */
```
send to the server ?

snalli replied:
or is this just a method used by a producer to put something into the queue ?

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+  /** Get the next request or block until there is one.
+   * @return*/
+  @Override
+  public NetworkRequestBundle receiveRequest() throws InterruptedException {
```
this receives requests from the client ?

snalli replied:
or is this just a method used by a consumer to get something frm the queue ?

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
```
i feel like this loop can be simplified. 

`  @Override
  public List<NetworkRequest> getUnqueuedRequestsX() throws InterruptedException {
    List<NetworkRequest> requestsToDrop = new ArrayList<>();
    NetworkRequest unqueuedRequest;
    while (requestsToDrop.isEmpty()) {
      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
        if (consumeStream(requestToDrop)) {
          requestsToDrop.add(unqueuedRequest);
        }
      }
      return requestsToDrop;
    }
  }`

snalli replied:
if requestsToDrop is empty, will a call to this method block ?

Arun-LinkedIn replied:
Thank you for suggestion. Simplified the loop.

> or why not let just dropRequest handle what to do with these errorneous requests instead of branching here ?

I think we need to keep code to read first 8 bytes at this layer so that it can be applicable to both valid and dropped requests. 

> if requestsToDrop is empty, will a call to this method block ?

Yeah, this method will block if requestsToDrop is empty. Added it as comments in javadoc.

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        if (!consumeStream(requestToDrop)) {
```
1/ why do we consumeStream if we have decided to drop ?
2/ and what do we do with the memory occupied by a request whose stream we cannot consume ? that doesn't seem to get added to the requeststoDrop queue ?

snalli replied:
can we explicitly release the mem for those erroneous requests ?

Arun-LinkedIn replied:
> 1/ why do we consumeStream if we have decided to drop ?

When we say drop, we actually mean to reject the reject without handling it. Due to that, we still need to consume the first 8 bytes of the stream. Added more comments in javadoc comments on why we are consuming the first 8 bytes.

> can we explicitly release the mem for those erroneous requests ?

Thank you for pointing it out. Released the memory for erroneous requests

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest unqueuedRequest;
+    do {
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
```
hope the poll is blocking else this would be an infinite spinning loop if the drop queue is always empty..

snalli commented on ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java:
```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+   * @param request incoming network request.
+   * @return false if there was any error while reading the first 8 bytes. Else, return true.
+   */
+  private boolean consumeStream(NetworkRequest request) throws InterruptedException {
```
this just seems like some verification method right ?

snalli replied:
better name would verifyXXXHeader ?

Arun-LinkedIn replied:
These are actually to consume/discard the first 8 bytes. Added few java docs comments and renamed it to `consumeSizeHeader`. Hopefully it is less confusing now.

snalli commented on ambry-network/src/main/java/com/github/ambry/network/LocalRequestResponseChannel.java:
```diff
@@ -45,13 +46,18 @@ public void sendRequest(NetworkRequest request) {
 
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
+  public NetworkRequestBundle receiveRequest() throws InterruptedException {
```
is the local channel in use ?

Arun-LinkedIn replied:
Yeah, local channel was actually used for Cloud requests. I will remove it in a separate pr.

snalli commented on ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java:
```diff
@@ -196,6 +196,77 @@ public void handleRequests(NetworkRequest networkRequest) throws InterruptedExce
+      switch (type) {
+        case PutRequest:
+          PutRequest putRequest = PutRequest.readFrom(dis, clusterMap);
+          putRequest.accept(metricsUpdater);
```
why not just metricsUpdater.accept here ? 
also, the accept doesn't really accept, kind of a misnomer

snalli commented on ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java:
```diff
@@ -196,6 +196,77 @@ public void handleRequests(NetworkRequest networkRequest) throws InterruptedExce
+          AdminRequest adminRequest = AdminRequest.readFrom(dis, clusterMap);
+          adminRequest.accept(metricsUpdater);
+          response = new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(),
+              ServerErrorCode.Retry_After_Backoff);
```
Retry_After_Backoff == 429 ?

Arun-LinkedIn replied:
Yeah, it is an error code understood by Ambry frontends. We can treat it as similar to 429 in frontends and retry the requests.

snalli commented on ambry-protocol/src/main/java/com/github/ambry/protocol/RequestDropper.java:
```diff
@@ -0,0 +1,69 @@
+        for (NetworkRequest requestToDrop : requestsToDrop) {
+          if (requestToDrop.equals(EmptyRequest.getInstance())) {
+            logger.debug("Request dropper received shut down command");
+            return;
```
where do we actually drop the request and release all the mem occupied by it ?

snalli commented on ambry-protocol/src/main/java/com/github/ambry/protocol/RequestDropper.java:
```diff
@@ -0,0 +1,69 @@
+    this.requests = requests;
+  }
+
+  public void run() {
```
can this thread get blocked on sending a response ? if so, we might be blocked for a while.

snalli commented on ambry-protocol/src/main/java/com/github/ambry/protocol/RequestDropper.java:
```diff
@@ -0,0 +1,69 @@
+    }
+  }
+
+  public void shutdown() throws InterruptedException {
```
can we have a more direct way of shutting down the thread ? like set a boolean and have the run method above just return when that boolean is true ?

snalli replied:
like while(shutdown == false)

snalli commented on ambry-api/src/main/java/com/github/ambry/config/NetworkConfig.java:
```diff
@@ -119,6 +122,35 @@ public class NetworkConfig {
+   * TODO: Do tests to tune controlled delay queue parameters before switching the request queue type to {@link RequestQueueType#ADAPTIVE_LIFO_CO_DEL}
+   */
+  @Config("request.queue.type")
+  @Default("BASIC_FIFO")
```
Nit: just fifo would do unless you have an advance fifo and a basic fifo

snalli replied:
Also these defaults are not useful anywhere

Arun-LinkedIn replied:
Actually meant to name as basic queue of type FIFO. Renamed it. Hopefully it is less confusing now.

snalli commented on ambry-api/src/main/java/com/github/ambry/config/NetworkConfig.java:
```diff
@@ -119,6 +122,35 @@ public class NetworkConfig {
+   * The maximum time in milliseconds a request is allowed to spend in the queue waiting to be processed.
+   */
+  @Config("request.queue.timeout.ms")
+  @Default("2000")
```
how do u decide these defaults ?

Arun-LinkedIn replied:
Discussed with @justinlin-linkedin offline on coming up with this defaults. It is related to our request timeout at frontends and request processing time at server. Since our frontends mostly have request time out of 4 seconds and if we assume request processing time around 1.5 second (based on P95 values in our graphs) and give another 0.5 seconds to network trip, we can probably allow 2 seconds for the requests in the queue before declaring them stale.

justinlin-linkedin reviewed (approved):
lgtm

tbingol reviewed (approved):
LGTM. Let's return a generic throttling error rather than a request-specific one in the following iterations

JingQianCloud reviewed (commented):
As we discussed, if we drop request whenever the 64 entry queue is full is probably too aggressive. We have 10 disks. So each disk has queue depth as 6. It's not very heavy.
Without retry mechanism in the frontend, we may return too many 503.

I think we shouldn't drop request only based on the queue size. We probably should queue many more requests but start to drop request when the response time or the queue time is too long.

snalli commented:
@Arun-LinkedIn Can you please address my comments before merging ?

Arun-LinkedIn commented on ambry-protocol/src/main/java/com/github/ambry/protocol/RequestDropper.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
```
Done.

mikeyang923 commented:
Can you address the unit test failure?
```
com.github.ambry.frontend.AmbrySecurityServiceTest > processResponseTest FAILED
    java.lang.AssertionError: Value of x-ambry-enc-um-filename not as expected expected:<%E6%B5%8B%E8%AF%95> but was:<null>

```

JingQianCloud reviewed (approved):
The unbounded queue solution which drops request only by timeout looks good to me.
